src/solver/conditions.rs

```diff
@@ -0,0 +1,124 @@
+//! Resolvo supports conditional dependencies. E.g. `foo REQUIRES bar IF baz is
+//! selected`.
+//!
+//! In SAT terms we express the requirement `A requires B` as `¬A1 v B1 .. v
+//! B99` where A1 is a candidate of package A and B1 through B99 are candidates
+//! that match the requirement B. In logical terms we say, either we do not
+//! select A1 or we select one of matching Bs.
+//!
+//! If we add a condition C to the requirement, e.g. `A requires B if C` we can
+//! modify the requirement clause to `¬A1 v ¬(C) v B1 .. v B2`. In logical terms
+//! we say, either we do not select A1 or we do not match the condition or we
+//! select one of the matching Bs.
+//!
+//! The condition `C` however expands to another set of matching candidates
+//! (e.g. `C1 v C2 v C3`). If we insert that into the formula we get,
+//!
+//!   `¬A1 v ¬(C1 v C2 v C3) v B1 .. v B2`
+//!
+//! which expands to
+//!
+//!   `¬A1 v ¬C1 ^ ¬C2 ^ ¬C3 v B1 .. v B2`
+//!
+//! This is not in CNF form (required for SAT clauses) so we cannot use this
+//! directly. To work around that problem we instead represent the condition
+//! `¬(C)` as the complement of the version set C. E.g. if C would represent
+//! `package >=1` then the complement would represent the candidates that match
+//! `package <1`, or the candidates that do NOT match C. So if we represent the
+//! complement of C as C! the final form of clause becomes:
+//!
+//!   `¬A1 v C!1 .. v C!99 v B1 .. v B2`
+//!
+//! This introduces another edge case though. What if the complement is empty?
+//! The final format would be void of `C!n` variables so it would become `¬A1 v
+//! B1 .. v B2`, e.g. A unconditionally requires B. To fix that issue we
+//! introduce an auxiliary variable that encodes if at-least-one of the package
+//! C is selected (notated as `C_selected`). For each candidate of C we add the
+//! clause
+//!
+//!   `¬Cn v C_selected`
+//!
+//! This forces `C_selected` to become true if any `Cn` is set to true. We then
+//! modify the requirement clause to be
+//!
+//!   `¬A1 v ¬C_selected v B1 .. v B2`
+//!
+//! Note that we do not encode any clauses to force `C_selected` to be false.
+//! We argue that this state is not required to function properly. If
+//! `C_selected` would be set to false it would mean that all candidates of
+//! package C are unselectable. This would disable the requirement, e.g. B
+//! shouldnt be selected for A1. But it doesnt prevent A1 from being selected.
+//!
+//! Conditions are expressed as boolean expression trees. Internally  they are
+//! converted to Disjunctive Normal Form (DNF). A boolean expression is
+//! in DNF if it is a **disjunction (OR)** of one or more **conjunctive clauses
+//! (AND)**.
```

Comment:
```suggestion
//! in DNF if the operands on the variables inside the clauses are **AND** while the clauses themselves are connected by **OR**'s
```

Reply:
I first interpreted the other as the other way around, this is my attempt to clarify while straying from the regular nomenclature. You decide!